File: windows/examples/f90examples/sd_unl.f
C     sd_unl.f
C       This program creates an HDF file called "sdunl.hdf"
C       and makes use of the unlimited dimension feature in
C       HDF.
C
C     
      PROGRAM UNLIMITED_DIM 

      implicit none

      integer*4 sd_id, sds_id, rank
      integer dims(2), start(2), edges(2), stride(2)
      integer i, j, istat
      integer sfstart, sfcreate, sfwdata, sfendacc, sfend
      integer sfselect

C     DFACC_CREATE and DFNT_INT16 are defined in hdf.h.
      integer*4 DFACC_CREATE, DFNT_INT16, DFACC_WRITE
      integer*4 DIM1, DIM0, SD_UNLIMITED, FAIL
      parameter (DFACC_CREATE = 4, DFNT_INT16 = 22, DIM1 = 5,
     +			 DIM0 = 10, DFACC_WRITE = 2, SD_UNLIMITED = 0,
     +            FAIL = -1)
      integer*2 array_data(DIM1, DIM0), append_data(DIM1)

C     Create and open the file and initiate the SD interface.
      sd_id = sfstart('sdunl.hdf', DFACC_CREATE)
      if (sd_id .ne. FAIL) then
         print*
         print*,'... sdunl.hdf created'
      end if

C     Define the rank and dimensions of the data set to be created.
      rank = 2
      dims(1) = DIM1
      dims(2) = SD_UNLIMITED

C     Create the data set.
      sds_id = sfcreate(sd_id, 'data1', DFNT_INT16, rank, dims)
      if (sds_id .ne. FAIL) then
        print*,'... data1 created'
      end if

C     Fill the stored-data array with values.
      do 20 j = 1, DIM0 
       do 10 i = 1, DIM1
        array_data(i, j) = i + j - 1
10     continue
20    continue

C     Define the location, pattern, and size of the data set
C     that will be written to.
      start(1) = 0
      start(2) = 0
      edges(1) = DIM1
      edges(2) = DIM0
      stride(1) = 1
      stride(2) = 1

C     Write the stored data to the "data1" data set.
      istat = sfwdata(sds_id, start, stride, edges, array_data)
      if (istat .ne. FAIL) then
         print*,'... data written out'
      end if

C     Terminate access to the array.
      istat = sfendacc(sds_id)

      istat = sfend (sd_id)
      if (istat .ne. FAIL) then
         print*,'... file closed'
      end if

C     Re-open the file and append data to the end of the file
      sd_id = sfstart('sdunl.hdf', DFACC_WRITE)
      if (sd_id .ne. FAIL) then
         print*,'... sdunl.hdf opened in Read/Write mode'
      end if

      sds_id = sfselect (sd_id, 0)
      if (sds_id .ne. FAIL) then
         print*,'... data1 selected'
      end if

      do 30 j = 1, DIM1
        append_data(j) = j 
30    continue

C     Define the location of the append
      start(1) = 0
      start(2) = DIM0
      edges(1) = DIM1
      edges(2) = 1

C     Append data to file
      istat = sfwdata(sds_id, start, stride, edges, append_data)
      if (istat .ne. FAIL) then
         print*,'... data appended'
      end if

C     Terminate access to the array.
      istat = sfendacc(sds_id)

C     Terminate access to the SD interface and close the file.
      istat = sfend(sd_id)
      if (istat .ne. FAIL) then
         print*,'... file closed'
         print*
      end if

      end

File: windows/examples/cexamples/sd_unl.c
/******************************************************************/
/*                                                                */
/*  sd_unl.c                                                      */
/*    This program creates an hdf file called "sdunl.hdf"         */
/*    and makes use of the unlimited dimension feature in         */
/*    HDF.                                                        */
/*                                                                */
/******************************************************************/

#include "mfhdf.h"

#define DIM0 10
#define DIM1 5

main( ) 
{
	int32 sd_id, sds_id, istat, sds_idx;
	int32 dims[2], start[2], edges[2], rank;
	int16 array_data[DIM0][DIM1], append_data[DIM1];
	intn i, j;

	/* Create and open the file and initiate the SD interface. */
	sd_id = SDstart("sdunl.hdf", DFACC_CREATE);
	if (sd_id != FAIL)
		printf ("\n... sdunl.hdf created\n");

	/* Define the rank and dimensions of the data set to be created. */
	rank = 2;
	dims[0] = SD_UNLIMITED;
	dims[1] = DIM1;

	/* Create the array data set. */
	sds_id = SDcreate(sd_id, "data1", DFNT_INT16, rank, dims);
	if (sds_id != FAIL)
		printf ("... data1 created\n");

	/* Fill the stored-data array with values. */
	for (j = 0; j < DIM0; j++) {
		for (i = 0; i < DIM1; i++)
			array_data[j][i] = (i + j) + 1;
 	}

	/* Define the location, pattern, and size of the data set */
	for (i = 0; i < rank; i++) {
		start[i] = 0;
	}

	edges[0] = DIM0;
	edges[1] = DIM1;

	istat = SDwritedata(sds_id, start, NULL, edges, (VOIDP)array_data); 
	if (istat != FAIL)
		printf ("... data written out\n");

	/* Terminate access to the array. */
	istat = SDendaccess(sds_id);

	/* Terminate access to the SD interface and close the file. */
	istat = SDend(sd_id);
	if (istat != FAIL)
		printf("... file closed\n");

	sd_id = SDstart("sdunl.hdf", DFACC_WRITE);
	if (sd_id != FAIL)
		printf ("... sdunl.hdf opened in Read/Write mode\n");

	sds_idx = SDnametoindex (sd_id, "data1");

	sds_id = SDselect (sd_id, sds_idx);
	if (sds_id != FAIL)
		printf ("... data1 selected\n");

	/* Store array values to be appended */
	for (i = 0; i <DIM1; i++)
		append_data[i] = i + 1;

	/* Define the location of the append */
	start[0] = DIM0;
	start[1] = 0;
	edges[0] = 1;
	edges[1] = DIM1;

	/* Append data to file */
	istat = SDwritedata (sds_id, start, NULL, edges, (VOIDP) append_data);
	if (istat != FAIL)
		printf ("... data appended\n");

	/* Terminate access to the array. */
	istat = SDendaccess (sds_id);

	/* Terminate access to the SD interface and close the file. */
	istat = SDend (sd_id);
	if (istat != FAIL)
		printf("... file closed\n\n");
}
